staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go

```diff
@@ -47,6 +49,34 @@ func init() {
 	utilfeature.DefaultMutableFeatureGate.Add(defaultKubernetesFeatureGates)
 }
 
+var (
+	CRDValidationRatcheting2 = featuregates.NewFeatureGate("CRDValidationRatcheting").
+					EnableByDefault().
+					Beta().
+					ToFeatureGateOrDie()
+
+	CustomResourceFieldSelectors2 = featuregates.NewFeatureGate("CustomResourceFieldSelectors").
+					Beta().
+					ToFeatureGateOrDie()
+
+	libraryFeatureSet = featuregates.NewSimpleFeatureSet()
+)
+
+func init() {
+	libraryFeatureSet.AddFeatureGatesOrDie(
+		CRDValidationRatcheting2,
+		CustomResourceFieldSelectors2,
+	)
+	libraryFeatureSet.AddFeatureSetsOrDie(
```

Comment:
notice composition here, where the apiextensions-apiserver needs client-go and can (or could choose not to) add the client-go featureset

Reply:
Do we think it's a good idea for libraries to make this choice (to include other libraries' gates)?  If the goal is composition and giving control to `main()`, isn't this counter to the goal?  What if `main()` doesn't want to expose the client-go gates?

This also applies to env-vars -- what if main doesn't want the end-user to be able to tweak gates?

Reply:
I think it's reasonable for a CRD-apiserver (or other server) to say, "I rely on the state of this FeatureGate in another library".  If we allow that, then I think allowing the composition is important since it becomes: "these are the featuregates important for my functionality".  Take the `WatchListClient` feature: is changing its default to true a kube-controller-manager main prerogative or something the `NewControllerManagerCommand() *cobra.Command ` should do so everywhere embedding that command gets the same behavior?

> what if main doesn't want the end-user to be able to tweak gates?
I'm less sure that I think we should encourage binaries to prevent libraries from reading env vars, but I suppose I could get behind it as an option.

Reply:
This feels like it is trending back to a global registry.  Is there ever a time where library A wants to use library B (both having gates), and library A does NOT want library B's gates included?

IOW if main uses A and exposes A's gates, they will get B's gates.  You're saying (I think) that is correct because A chose so.  When would A not choose so?

When would main NOT want the union of all the gates linked into it?